File: pkg/models/final_output_specs.go
package models

import "encoding/json"

const FinalOutputSpecVersion = "1"

type DocumentSpec struct {
	Version  string             `json:"version"`
	Title    string             `json:"title"`
	Subtitle string             `json:"subtitle,omitempty"`
	Metadata []DocumentMetadata `json:"metadata,omitempty"`
	Sections []DocumentSection  `json:"sections"`
}

type DocumentMetadata struct {
	Label string `json:"label"`
	Value string `json:"value"`
}

type DocumentSection struct {
	Title  string          `json:"title"`
	Blocks []DocumentBlock `json:"blocks"`
}

type DocumentBlock struct {
	Type  string         `json:"type"`
	Text  string         `json:"text,omitempty"`
	Title string         `json:"title,omitempty"`
	Tone  string         `json:"tone,omitempty"`
	Items []string       `json:"items,omitempty"`
	Table *DocumentTable `json:"table,omitempty"`
}

type DocumentTable struct {
	Columns []string   `json:"columns"`
	Rows    [][]string `json:"rows"`
}

type SpreadsheetSpec struct {
	Version string             `json:"version"`
	Title   string             `json:"title,omitempty"`
	Sheets  []SpreadsheetSheet `json:"sheets"`
}

type SpreadsheetSheet struct {
	Name         string                       `json:"name"`
	Columns      []SpreadsheetColumn          `json:"columns"`
	Rows         []map[string]json.RawMessage `json:"rows"`
	FreezeHeader bool                         `json:"freeze_header,omitempty"`
	AutoFilter   bool                         `json:"auto_filter,omitempty"`
}

type SpreadsheetColumn struct {
	Key          string  `json:"key"`
	Header       string  `json:"header"`
	Width        float64 `json:"width,omitempty"`
	NumberFormat string  `json:"number_format,omitempty"`
}

type DashboardSpec struct {
	Version string           `json:"version"`
	Title   string           `json:"title"`
	Blocks  []DashboardBlock `json:"blocks"`
}

type DashboardBlock struct {
	Type     string                       `json:"type"`
	Title    string                       `json:"title,omitempty"`
	Text     string                       `json:"text,omitempty"`
	Tone     string                       `json:"tone,omitempty"`
	Status   string                       `json:"status,omitempty"`
	Label    string                       `json:"label,omitempty"`
	Value    string                       `json:"value,omitempty"`
	Href     string                       `json:"href,omitempty"`
	Items    []DashboardBlockItem         `json:"items,omitempty"`
	Columns  []DashboardTableColumn       `json:"columns,omitempty"`
	Rows     []map[string]json.RawMessage `json:"rows,omitempty"`
	Progress *DashboardProgress           `json:"progress,omitempty"`
	Chart    *DashboardChart              `json:"chart,omitempty"`
}

type DashboardBlockItem struct {
	Label  string `json:"label,omitempty"`
	Value  string `json:"value,omitempty"`
	Text   string `json:"text,omitempty"`
	Status string `json:"status,omitempty"`
	Tone   string `json:"tone,omitempty"`
	Href   string `json:"href,omitempty"`
}

type DashboardTableColumn struct {
	Key   string `json:"key"`
	Label string `json:"label"`
}

type DashboardProgress struct {
	Value float64 `json:"value"`
	Max   float64 `json:"max,omitempty"`
	Unit  string  `json:"unit,omitempty"`
}

type DashboardChart struct {
	Type                string                 `json:"type"`
	Unit                string                 `json:"unit,omitempty"`
	TimeWindow          *DashboardTimeWindow   `json:"time_window,omitempty"`
	AggregationInterval string                 `json:"aggregation_interval,omitempty"`
	MissingValues       string                 `json:"missing_values,omitempty"`
	Dimensions          map[string]string      `json:"dimensions,omitempty"`
	Series              []DashboardChartSeries `json:"series"`
}

type DashboardTimeWindow struct {
	From string `json:"from,omitempty"`
	To   string `json:"to,omitempty"`
}

type DashboardChartSeries struct {
	Key         string                 `json:"key"`
	Label       string                 `json:"label,omitempty"`
	Team        string                 `json:"team,omitempty"`
	Environment string                 `json:"environment,omitempty"`
	Unit        string                 `json:"unit,omitempty"`
	Color       string                 `json:"color,omitempty"`
	Points      []DashboardSeriesPoint `json:"points"`
}

type DashboardSeriesPoint struct {
	Timestamp string   `json:"timestamp,omitempty"`
	Label     string   `json:"label,omitempty"`
	Value     *float64 `json:"value,omitempty"`
}

File: services/ui/src/features/pipeline-runs/final-output-preview/finalOutputSpecs.ts
export type DocumentSpec = {
  version: '1';
  title: string;
  subtitle?: string;
  metadata?: Array<{ label: string; value: string }>;
  sections: Array<{
    title: string;
    blocks: DocumentBlock[];
  }>;
};

export type DocumentBlock =
  | { type: 'paragraph'; text: string }
  | { type: 'bullet_list' | 'numbered_list'; items: string[] }
  | { type: 'table'; table: { columns: string[]; rows: string[][] } }
  | { type: 'callout'; title?: string; tone?: 'info' | 'success' | 'warning' | 'critical'; text: string };

export type SpreadsheetSpec = {
  version: '1';
  title?: string;
  sheets: Array<{
    name: string;
    columns: Array<{
      key: string;
      header: string;
      width?: number;
      number_format?: string;
    }>;
    rows: Array<Record<string, string | number | boolean | null>>;
    freeze_header?: boolean;
    auto_filter?: boolean;
  }>;
};

export function parseDocumentSpec(content: string): DocumentSpec | null {
  const value = parseObject(content);
  if (!value || value.version !== '1' || typeof value.title !== 'string' || !Array.isArray(value.sections)) return null;
  return value as DocumentSpec;
}

export function parseSpreadsheetSpec(content: string): SpreadsheetSpec | null {
  const value = parseObject(content);
  if (!value || value.version !== '1' || !Array.isArray(value.sheets)) return null;
  return value as SpreadsheetSpec;
}

export function documentSpecToText(spec: DocumentSpec) {
  const lines = [spec.title];
  if (spec.subtitle) lines.push(spec.subtitle);
  for (const item of spec.metadata || []) lines.push(`${item.label}: ${item.value}`);
  for (const section of spec.sections) {
    lines.push('', section.title);
    for (const block of section.blocks) {
      if (block.type === 'paragraph' || block.type === 'callout') lines.push(block.text);
      if (block.type === 'bullet_list') lines.push(...block.items.map(item => `- ${item}`));
      if (block.type === 'numbered_list') lines.push(...block.items.map((item, index) => `${index + 1}. ${item}`));
      if (block.type === 'table') {
        lines.push(block.table.columns.join('\t'));
        lines.push(...block.table.rows.map(row => row.join('\t')));
      }
    }
  }
  return lines.join('\n').trim();
}

export function spreadsheetSpecToText(spec: SpreadsheetSpec) {
  const sheets = spec.sheets.map(sheet => {
    const rows = [sheet.columns.map(column => column.header).join('\t')];
    for (const row of sheet.rows) {
      rows.push(sheet.columns.map(column => formatCell(row[column.key])).join('\t'));
    }
    return [`[${sheet.name}]`, ...rows].join('\n');
  });
  return sheets.join('\n\n');
}

export function formatCell(value: unknown) {
  if (value === null || value === undefined) return '';
  if (typeof value === 'boolean') return value ? 'Yes' : 'No';
  return String(value);
}

function parseObject(content: string): Record<string, unknown> | null {
  try {
    const value: unknown = JSON.parse(content);
    return value !== null && typeof value === 'object' && !Array.isArray(value) ? (value as Record<string, unknown>) : null;
  } catch {
    return null;
  }
}
